File: app/utils/password.py
"""
Password generator untuk temporary password.

Dijamin mengandung:
- minimal 1 huruf besar
- minimal 1 huruf kecil
- minimal 1 angka
- minimal 1 karakter spesial

Sehingga lulus password policy default FreeIPA.
"""

import secrets
import string


def generate_temporary_password(length: int = 16) -> str:
    """
    Generate password random yang aman secara kriptografis.

    Pakai `secrets` (bukan `random`) karena cocok untuk hal sensitif
    seperti password / token.
    """
    if length < 8:
        raise ValueError("Panjang password minimal 8 karakter")

    upper = string.ascii_uppercase
    lower = string.ascii_lowercase
    digits = string.digits
    # Hindari karakter yang ambigu / bermasalah di shell, email, dsb.
    specials = "!@#$%^&*-_=+"

    # Pastikan minimal 1 karakter dari setiap kategori
    required = [
        secrets.choice(upper),
        secrets.choice(lower),
        secrets.choice(digits),
        secrets.choice(specials),
    ]

    # Sisanya random dari semua kategori
    all_chars = upper + lower + digits + specials
    remaining = [secrets.choice(all_chars) for _ in range(length - len(required))]

    # Acak posisinya supaya karakter wajib tidak selalu di awal
    password_chars = required + remaining
    secrets.SystemRandom().shuffle(password_chars)

    return "".join(password_chars)

File: backend/internal/utils/password.go
// Package utils berisi helper-helper kecil yang reusable.
package utils

import (
	"crypto/rand"
	"crypto/subtle"
	"errors"
	"math/big"
)

const (
	upperChars   = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	lowerChars   = "abcdefghijklmnopqrstuvwxyz"
	digitChars   = "0123456789"
	specialChars = "!@#$%^&*-_=+"
)

// GenerateTemporaryPassword menghasilkan password acak yang
// kriptografis-aman, dijamin mengandung minimal 1 huruf besar,
// 1 huruf kecil, 1 angka, dan 1 karakter spesial.
func GenerateTemporaryPassword(length int) (string, error) {
	if length < 12 {
		return "", errors.New("panjang password minimal 12 karakter")
	}

	all := upperChars + lowerChars + digitChars + specialChars

	// Karakter wajib (1 dari setiap kategori)
	required := []byte{
		mustPick(upperChars),
		mustPick(lowerChars),
		mustPick(digitChars),
		mustPick(specialChars),
	}

	// Sisanya random dari semua kategori
	out := make([]byte, length)
	copy(out, required)
	for i := len(required); i < length; i++ {
		out[i] = mustPick(all)
	}

	// Acak posisi (Fisher-Yates) supaya karakter wajib tidak selalu di awal
	for i := length - 1; i > 0; i-- {
		j := mustIntn(i + 1)
		out[i], out[j] = out[j], out[i]
	}

	return string(out), nil
}

// ConstantTimeEqual membandingkan 2 string dalam waktu konstan
// untuk mencegah timing attack (mis. saat membandingkan token).
func ConstantTimeEqual(a, b string) bool {
	return subtle.ConstantTimeCompare([]byte(a), []byte(b)) == 1
}

// MaskEmail mengubah "john.doe@example.com" -> "jo***@example.com".
// Dipakai di response supaya email user lain tidak bocor.
func MaskEmail(email string) string {
	at := -1
	for i, c := range email {
		if c == '@' {
			at = i
			break
		}
	}
	if at <= 0 {
		return "***"
	}
	local := email[:at]
	domain := email[at:]
	if len(local) <= 2 {
		return string(local[0]) + "***" + domain
	}
	return local[:2] + "***" + domain
}

func mustPick(charset string) byte {
	idx := mustIntn(len(charset))
	return charset[idx]
}

func mustIntn(n int) int {
	if n <= 0 {
		return 0
	}
	v, err := rand.Int(rand.Reader, big.NewInt(int64(n)))
	if err != nil {
		// Sangat jarang terjadi; kalau crypto/rand gagal, sistem bermasalah.
		panic("crypto/rand failed: " + err.Error())
	}
	return int(v.Int64())
}
